package com.ferra13671.BThack.impl.modules.render;

import com.ferra13671.BThack.api.module.ModuleInfo;
import com.ferra13671.BThack.core.render.BThackRender;
import com.ferra13671.BThack.core.render.box.RenderBox;
import com.ferra13671.BThack.events.block.UseBlockEvent;
import com.ferra13671.BThack.events.GuiOpenEvent;
import com.ferra13671.BThack.events.render.RenderWorldLastEvent;
import com.ferra13671.BThack.events.ClientTickEvent;
import com.ferra13671.BThack.managers.impl.setting.Settings.ColorSetting;
import com.ferra13671.BThack.api.module.Module;
import com.ferra13671.BThack.api.utils.BlockUtils;
import com.ferra13671.BThack.api.utils.MathUtils;
import com.ferra13671.MegaEvents.Base.EventSubscriber;
import net.minecraft.block.Block;
import net.minecraft.block.Blocks;
import net.minecraft.block.ShulkerBoxBlock;
import net.minecraft.client.gui.screen.ingame.GenericContainerScreen;
import net.minecraft.client.gui.screen.ingame.ShulkerBoxScreen;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Box;
import net.minecraft.util.math.Vec3d;

import java.awt.*;
import java.util.ArrayList;
import java.util.List;

@ModuleInfo(name = "LastOpenChest", description = "lang.module.LastOpenChest", category = "RENDER")
public class LastOpenChest extends Module {

    public final ColorSetting colorSet = new ColorSetting("Color", this, new Color(0, 255, 255)).withBlockedAlpha();


    private BlockPos chestPos;
    private boolean needRender = false;

    @EventSubscriber
    @SuppressWarnings("unused")
    public void onUseBlock(UseBlockEvent e) {
        if (isChest(e.getBlockHitResult().getBlockPos())) {
            chestPos = e.getBlockHitResult().getBlockPos();
            needRender = false;
        }
    }


    @EventSubscriber
    @SuppressWarnings({"unused", "DataFlowIssue"})
    public void onGuiOpen(GuiOpenEvent e) {
        if (e.getScreen() instanceof GenericContainerScreen || e.getScreen() instanceof ShulkerBoxScreen) {
            if (chestPos != null) {
                if (MathUtils.getDistance(mc.player.getPos(), new Vec3d(chestPos.getX(), chestPos.getY(), chestPos.getZ())) > 20) {
                    chestPos = null;
                    needRender = false;
                } else
                    needRender = true;
            }
        }
    }

    @EventSubscriber
    @SuppressWarnings("unused")
    public void onTick(ClientTickEvent e) {
        if (nullCheck()) return;

        if (chestPos != null && !isChest(chestPos))
            chestPos = null;
    }

    @EventSubscriber
    @SuppressWarnings("unused")
    public void onRender(RenderWorldLastEvent e) {
        if (chestPos != null && needRender) {
            float _red = colorSet.getValue().getRed() / 255f;
            float _green = colorSet.getValue().getGreen() / 255f;
            float _blue = colorSet.getValue().getBlue() / 255f;

            Box box = BlockUtils.getBoundingBox(chestPos);
            if (box == null) return;

            BThackRender.boxRender.prepareBoxRender();
            BThackRender.boxRender.renderBoxes(new ArrayList<>(List.of(new RenderBox(box, _red, _green, _blue, 0.6f, _red, _green, _blue, 0.4f))));
            BThackRender.boxRender.stopBoxRender();
        }
    }

    @SuppressWarnings("DataFlowIssue")
    private boolean isChest(BlockPos pos) {
        Block block = mc.world.getBlockState(pos).getBlock();
        return block == Blocks.CHEST || block == Blocks.ENDER_CHEST || block == Blocks.TRAPPED_CHEST || block == Blocks.BARREL || block instanceof ShulkerBoxBlock;
    }
}
